src/main/thrift/parquet.thrift

```diff
@@ -885,6 +971,44 @@ struct ColumnChunk {
   9: optional binary encrypted_column_metadata
 }
 
+struct ColumnChunkV3 {
+  /** File where column data is stored. **/
+  1: optional string file_path
```

Comment:
I think we do support it in PyArrow (see tests [here](https://github.com/apache/arrow/blob/1c546fb3c130fc6a4f3e06ad31dc49d923785104/python/pyarrow/tests/parquet/test_metadata.py#L459-L499) and [here](https://github.com/apache/arrow/blob/1c546fb3c130fc6a4f3e06ad31dc49d923785104/python/pyarrow/tests/test_dataset.py#L3614-L3638)), and I think [Dask](https://docs.dask.org/) makes use of it.

@jorisvandenbossche @mrocklin Am I right?

Reply:
This should be the summary metadata file, IIUC. cc @rdblue for more context.

Reply:
Deferring this question from Dask's perspective to @fjetter and @phofl

Reply:
(thanks for pinging us)

Reply:
@wgtmac mentioned keeping this feature to allow for "column families" (i.e. splitting columns across different files).

Reply:
Also, CC @sunchao

Reply:
Dask is indeed supporting this but it's causing quite some headaches. It's been the default for a while but it turned out to be impractical for any real world datasets and we are currently mostly supporting it for backwards compatibility. Especially when including statistics, this became just impractical to use due to it's size so this new metadata proposal will likely not change that much about it's usefulness.

I can see how this could be used to build something useful but the way dask is using it right now is not reason enough to keep this. (cc @rjzamora in case you have a different perspective).

Also, I'm not even sure if pyarrow exposes an API to store the metadata exclusively in an external file. From what I understand this is always a duplicate so this field may actually not be required to support what we are doing.

Reply:
I think we can remove this. There should be better approaches to support "column families".

Reply:
+1 to keeping metadata in the same file.

Reply:
Correct me if I am wrong, but the main purpose of this is to gather the row group metadata into a single file, so that when reading a large dataset, you don't need to open each and every file in order to read its metadata to decide whether any groups in that file need reading.

This is particularly expensive on networked filesystems, for very large datasets, so this feature is very useful for minimising io and latency in reading.

My usecase is reading the same large dataset from a compute farm (thousands of machines), and walking the directory tree and opening each file to read metadata from each of the farm nodes really upsets the network filesystem.

I guess I fail to see how this is not useful, but I'm happy for this to go as long as there is an alternative for this, which reduces io in the same way.

Reply:
@AudriusButkevicius a better place to argue this would be [here](https://docs.google.com/document/d/1PQpY418LkIDHMFYCY8ne_G-CFpThK15LLpzWYbc7rFU/edit#heading=h.cu6a3mbixmk2).

Reply:
That specific proposal does not discuss deprecating this field.

Reply:
We at G-Research would very much like to see support for `_metadata` files (and therefore the `file_path` field) remain in Parquet V3. Using a `_metadata` file provides significant performance benefit for us by reducing the number of file system operations required for certain queries when using the Arrow Dataset library. This is especially important when using a network file system. In a benchmark comparing query performance with and without using a `_metadata` file I found that queries were nearly 5x faster when using a Dataset built from the `_metadata` file. This was a [synthetic benchmark](https://github.com/adamreeve/dataset-metadata-benchmark/tree/caac9d5115cfcda309939cf7deafb78f36b41d37) using a local instance of minio rather than using our production data and infrastructure,  but demonstrates the performance benefit this `_metadata` file can provide.

We understand that the main reason for deprecating this field is that people can use alternative open table formats like Apache Iceberg or Delta Lake which provide similar performance benefits, but these add a lot of complexity and include many extra features that we don't always need. We also use .NET and are working on creating bindings to the Arrow Dataset library for .NET, and don't want to also have to develop a .NET Iceberg implementation.

There also doesn't seem to be much harm in keeping this field from what I can see, as it can safely be ignored by Parquet readers that don't use `_metadata` files.

Reply:
@adamreeve Interesting, you're the first person I have ever heard who would use this external file feature. 

However, the field does not seem to fully match what you are describing. The `file_path` field defines where the *data* itself lives, not the metadata. Or are you saying, that the "main" parquet file is your `_metadata` file and then the actual data lives in other files?

Can you elaborate what advantage you see here? I can't grasp how putting data into a different file would help with network file systems. Our system (and probably the majority on the industry) runs on cloud object stores and there having data in separate files wouldn't help at all. So how does it help in your use case? How does it reduce the number of system operations?

Reply:
The _metadata file has all of the row group details/stats. If I have a dataset that I need to read and filter to a given date (where date is not part of the filesystem partitioning scheme), I can use _metadata file to filter down the row groups, see which files those row groups belong to and only read those files. Otherwise I'd have to open every file, read it's row group stats, decide the file doesn't have the date I'm after, close file, move to the next file, repeat.

Reply:
I think Audrius has answered your question @JFinis (Audrius and I are work colleagues), I just want to add that I don't think what we're doing is particularly unusual as this is a feature of the Arrow Dataset library, so presumably there are other users of this API (see [pyarrow.dataset.parquet_dataset](https://arrow.apache.org/docs/python/generated/pyarrow.dataset.parquet_dataset.html)).

And the benefits of using this `_metadata` index file should translate to cloud object stores too by reducing the number of objects/files to be read.

Reply:
@adamreeve / @AudriusButkevicius do I understand correctly that you have a dataset with N parquet files and in your schema you have a `_metadata` column which contains metadata for all N files. Each of the files says data for `_metadata` column is this other file over here (`file_path`) so while logically each file has information for all other files, the data exists only in single file containing the metadata for the whole dataset.

Reply:
@alkis, no, there is no `_metadata` column in the schema. There is a file named `_metadata` which contains a copy of the metadata from all N Parquet files, with the only difference being that the metadata in this file has the `file_path` field set to the path of the file containing the data corresponding to each metadata copy. This `_metadata` file contains no data pages itself, but can be used like an index to determine which file to read data from based on the metadata.

Reply:
PyArrow provides [write_metadata])https://arrow.apache.org/docs/python/generated/pyarrow.parquet.write_metadata.html) and [parquet_dataset](https://arrow.apache.org/docs/python/generated/pyarrow.dataset.parquet_dataset.html#pyarrow.dataset.parquet_dataset) for such use case.

Reply:
> And the benefits of using this _metadata index file should translate to cloud object stores too by reducing the number of objects/files to be read. 

not really
1. forces a full read of all generated files in job commit, which even if done in parallel is really slow. If it were to be done, it'd be better off done on demand in the first query. (note, faster reads would improve this)
1. it doesn't work with the cloud committer design, which #1361 formalises without doing some bridging classes. 

the reason for (2) is that the hadoop cloud-native committer design kept clear of making any changes to the superclass of `ParquetOutputCommitter` as it is a critical piece of code in so many existing workflows, and really hard to understand. Not just a co-recursive algorithm, but two intermingled algorithms, one of which lacks the correctness guarantees (failures during task commit can be recovered from).

with a move to table based formats rather than directory trees, that whole commit process becomes much easier as well as supporting atomic job commits on a table (including deletes!). And as you note, these formats can include schema info too.

Reply:
For context, AFAIK the `_metadata` summary file was a practice originally used in Spark (and supported by Parquet-mr), and inspired on that for example also taken over in Dask. We then implemented support for this in Arrow C++ / PyArrow mostly based on the dask usage (as a downstream user of pyarrow). In the meantime though, Spark disabled writing those files by default a long time ago, and also dask stopped doing this 2 years ago (https://github.com/dask/dask/issues/8901),

Another Parquet dev mailing list thread with some discussion about this: https://lists.apache.org/thread/142yj57c68s2ob5wkrs80xsjoksm7rb7

Reply:
@adamreeve I see so the parquet file is one with all the metadata and all the data is in files pointed to by this singleton.

Reply:
>> And the benefits of using this _metadata index file should translate to cloud object stores too by reducing the number of objects/files to be read.

> not really

Sorry I couldn't really follow this argument, that sounds like a Hadoop specific problem. To me a cloud object store means something like S3, and for our use case we're mostly concerned with reducing the number of objects that need to be read to satisfy read queries that filter data and don't need to read all files in a dataset, as we have many concurrent jobs running and adding load on storage.

> Another Parquet dev mailing list thread with some discussion about this: https://lists.apache.org/thread/142yj57c68s2ob5wkrs80xsjoksm7rb7

Much of the discussion there seems to be related to issues users can run into if doing things like overwriting Parquet files or having heterogeneous schemas, which this feature was not designed for. But it sounds like others have also found this feature useful. I think this quote from Patrick Woody matches our experience: "outstandingly useful when you have well laid out data with a sort-order"

> @adamreeve I see so the parquet file is one with all the metadata and all the data is in files pointed to by this singleton.

Yes, exactly.

The `_metadata` file format could have been designed so that the `file_path` field wasn't needed in the column chunk metadata. But it's there now and provides value to users while adding minimal overhead to those not using it (missing fields require zero space in serialized messages if I've understood the [Thrift Compact Protocol](https://github.com/apache/thrift/blob/master/doc/specs/thrift-compact-protocol.md) correctly).